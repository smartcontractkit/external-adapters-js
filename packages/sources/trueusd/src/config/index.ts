import { AdapterConfig } from '@chainlink/external-adapter-framework/config'

export const config = new AdapterConfig({
  API_ENDPOINT: {
    description: 'API endpoint of adapter',
    type: 'string',
    default: 'https://api.real-time-reserves.ledgerlens.io/v1/',
  },
})
