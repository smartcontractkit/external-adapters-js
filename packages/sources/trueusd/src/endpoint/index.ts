export * as trueusd from './trueusd'
