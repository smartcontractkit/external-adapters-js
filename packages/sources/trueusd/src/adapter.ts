import { Requester, Validator, AdapterError } from '@chainlink/ea-bootstrap'
import { Config, ExecuteWithConfig, ExecuteFactory } from '@chainlink/types'
import { makeConfig, DEFAULT_ENDPOINT } from './config'
import { trueusd } from './endpoint'

const inputParams = {
  endpoint: false,
}

export const execute: ExecuteWithConfig<Config> = async (request, context, config) => {
  const validator = new Validator(request, inputParams)
  if (validator.error) throw validator.error

  Requester.logConfig(config)

  const jobRunID = validator.validated.id
  const endpoint = validator.validated.data.endpoint || DEFAULT_ENDPOINT

  switch (endpoint.toLowerCase()) {
    case trueusd.NAME: {
      return await trueusd.execute(request, context, config)
    }
    default: {
      throw new AdapterError({
        jobRunID,
        message: `Endpoint ${endpoint} not supported.`,
        statusCode: 400,
      })
    }
  }
}

export const makeExecute: ExecuteFactory<Config> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}
