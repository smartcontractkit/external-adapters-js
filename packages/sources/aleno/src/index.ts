import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { price } from './endpoint'

export const adapter = new Adapter({
  defaultEndpoint: price.name,
  name: 'ALENO',
  config,
  endpoints: [price],
  rateLimiting: {
    tiers: {
      default: {
        rateLimit1m: 30,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
