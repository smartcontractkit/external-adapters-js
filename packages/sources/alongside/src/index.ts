import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { collateral } from './endpoint'
import { config } from './config'

export const adapter = new Adapter({
  defaultEndpoint: collateral.name,
  name: 'ALONGSIDE',
  config,
  endpoints: [collateral],
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
