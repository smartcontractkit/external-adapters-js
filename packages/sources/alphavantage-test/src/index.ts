import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { PriceAdapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { forex } from './endpoint'

export const adapter = new PriceAdapter({
  defaultEndpoint: forex.name,
  name: 'ALPHAVANTAGE',
  config,
  endpoints: [forex],
  rateLimiting: {
    tiers: {
      free: {
        rateLimit1m: 5,
        rateLimit1h: 20.83,
      },
      '49.99': {
        rateLimit1m: 75,
      },
      '99.99': {
        rateLimit1m: 150,
      },
      '149.99': {
        rateLimit1m: 300,
      },
      '199.99': {
        rateLimit1m: 600,
      },
      '249.99': {
        rateLimit1m: 1200,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
