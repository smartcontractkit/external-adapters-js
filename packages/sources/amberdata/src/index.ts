import { expose } from '@chainlink/ea-bootstrap'
import { makeExecute } from './adapter'
import { makeConfig, WSHandlerFactory, NAME } from './config'

export = { NAME, makeExecute, makeConfig, ...expose(makeExecute(), WSHandlerFactory()) }
