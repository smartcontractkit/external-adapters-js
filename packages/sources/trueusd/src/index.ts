import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { PoRAdapter } from '@chainlink/external-adapter-framework/adapter/por'
import { trueusd } from './endpoint'
import { config } from './config'

export const adapter = new PoRAdapter({
  defaultEndpoint: trueusd.name,
  name: 'TRUEUSD',
  config,
  endpoints: [trueusd],
  rateLimiting: {
    tiers: {
      default: {
        rateLimit1m: 6,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
