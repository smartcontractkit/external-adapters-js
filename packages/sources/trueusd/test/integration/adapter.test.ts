import { AdapterRequest } from '@chainlink/ea-bootstrap'
import { server as startServer } from '../../src'
import { mockResponseSuccess } from './fixtures'
import { setupExternalAdapterTest } from '@chainlink/ea-test-helpers'
import type { SuiteContext } from '@chainlink/ea-test-helpers'
import { SuperTest, Test } from 'supertest'

describe('execute', () => {
  const id = '1'
  const context: SuiteContext = {
    req: null,
    server: startServer,
  }

  const envVariables = {
    CACHE_ENABLED: 'false',
  }

  setupExternalAdapterTest(envVariables, context)

  it('should return success for trust', async () => {
    const data: AdapterRequest = {
      id,
      data: {
        resultPath: 'totalTrust',
      },
    }

    mockResponseSuccess()

    const response = await (context.req as SuperTest<Test>)
      .post('/')
      .send(data)
      .set('Accept', '*/*')
      .set('Content-Type', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200)
    expect(response.body).toMatchSnapshot()
  })

  it('should return success for token', async () => {
    const data: AdapterRequest = {
      id,
      data: {
        resultPath: 'totalToken',
      },
    }

    mockResponseSuccess()

    const response = await (context.req as SuperTest<Test>)
      .post('/')
      .send(data)
      .set('Accept', '*/*')
      .set('Content-Type', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200)
    expect(response.body).toMatchSnapshot()
  })

  it('should return success when given a chain', async () => {
    const data: AdapterRequest = {
      id,
      data: {
        chain: 'AVA',
      },
    }

    mockResponseSuccess()

    const response = await (context.req as SuperTest<Test>)
      .post('/')
      .send(data)
      .set('Accept', '*/*')
      .set('Content-Type', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200)
    expect(response.body).toMatchSnapshot()
  })

  it('should return success when given a chain and resultPath', async () => {
    const data: AdapterRequest = {
      id,
      data: {
        chain: 'TUSD (AVAX)',
        resultPath: 'totalTokenByChain',
      },
    }

    mockResponseSuccess()

    const response = await (context.req as SuperTest<Test>)
      .post('/')
      .send(data)
      .set('Accept', '*/*')
      .set('Content-Type', 'application/json')
      .expect('Content-Type', /json/)
    expect(response.body).toMatchSnapshot()
  })
})
