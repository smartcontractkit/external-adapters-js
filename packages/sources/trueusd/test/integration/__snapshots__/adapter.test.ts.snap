// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute trueusd endpoint should return error when ripcord true 1`] = `
{
  "errorMessage": "Ripcord indicator true. Details: Integrations, Balances",
  "ripcord": true,
  "ripcordDetails": "Integrations, Balances",
  "statusCode": 502,
  "timestamps": {
    "providerDataReceivedUnixMs": 978347471111,
    "providerDataRequestedUnixMs": 978347471111,
    "providerIndicatedTimeUnixMs": 1686848387538,
  },
}
`;

exports[`execute trueusd endpoint should return success for token 1`] = `
{
  "data": {
    "result": 3097573780.109635,
    "ripcord": false,
  },
  "result": 3097573780.109635,
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 978347471111,
    "providerDataRequestedUnixMs": 978347471111,
    "providerIndicatedTimeUnixMs": 1686848387538,
  },
}
`;

exports[`execute trueusd endpoint should return success trust 1`] = `
{
  "data": {
    "result": 3117410009.89,
    "ripcord": false,
  },
  "result": 3117410009.89,
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 978347471111,
    "providerDataRequestedUnixMs": 978347471111,
    "providerIndicatedTimeUnixMs": 1686848387538,
  },
}
`;

exports[`execute trueusd endpoint should return success when given a chain 1`] = `
{
  "data": {
    "result": 24268.83,
    "ripcord": false,
  },
  "result": 24268.83,
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 978347471111,
    "providerDataRequestedUnixMs": 978347471111,
    "providerIndicatedTimeUnixMs": 1686848387538,
  },
}
`;

exports[`execute trueusd endpoint should return success when given a chain and field 1`] = `
{
  "data": {
    "result": 3005274.95,
    "ripcord": false,
  },
  "result": 3005274.95,
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 978347471111,
    "providerDataRequestedUnixMs": 978347471111,
    "providerIndicatedTimeUnixMs": 1686848387538,
  },
}
`;
